### PR TITLE
add spp（Spatial pyramid pooling ） op 

### PR DESCRIPTION
**Pull request by sweetsky0901**
#5988

### CODE REVIEW

File: paddle/operators/spp_op.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        "(Tensor) The input tensor of spp operator. "
+        "The format of input tensor is NCHW. Where N is batch size, C is the "
+        "number of channels, H and W is the height and width of feature.");
+    AddOutput("Out",
+              "(Tensor) The output tensor of spp operator."
+              "N * M."
+              "M = C * H * W");
+    AddAttr<int>("pyramid_height", "(int), multi level pooling");
+    AddComment(R"DOC(
+        "Does spatial pyramid pooling on the input image by taking the max,
```
Comment (NHZlX) [inline paddle/operators/spp_op.cc]: The document is too simple to be easy to understand for the novice.

Comment (NHZlX) [reply paddle/operators/spp_op.cc]: refer here https://github.com/tensorflow/tensorflow/issues/6011

File: paddle/operators/spp_op.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      int padding_w = (kernel_size_w * bins - input_w + 1) / 2;
+      std::vector<int> kernel_size({kernel_size_h, kernel_size_w});
+      std::vector<int> strides({kernel_size_h, kernel_size_w});
+      std::vector<int> paddings({padding_h, padding_w});
+      // pooling output shape
+      framework::Tensor out_level;
+      std::vector<int64_t> output_shape_vec({in_x->dims()[0], in_x->dims()[1]});
+      output_shape_vec.push_back(
+          (input_h - kernel_size_h + 2 * padding_h) / kernel_size_h + 1);
+      output_shape_vec.push_back(
```
Comment (NHZlX) [inline paddle/operators/spp_op.h]: Is `(input_h - kernel_size_h + 2 * padding_h) / kernel_size_h + 1)` equals to `bins`? Why do we have to recaulate.

File: paddle/operators/spp_op.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                   paddings, max_process, &out_level);
+      // flatten pooling output shape
+      framework::Tensor out_flatten_level;
+      int output_flatten_w = in_x->dims()[1] * bins * bins;
+      std::vector<int64_t> output_flatten_shape_vec(
+          {in_x->dims()[0], output_flatten_w});
+      framework::DDim output_flatten_shape(
+          framework::make_ddim(output_flatten_shape_vec));
+      out_flatten_level.ShareDataWith(out_level);
+      out_flatten_level.Resize(output_flatten_shape);
```
Comment (NHZlX) [inline paddle/operators/spp_op.h]: why not use o`ut_level` to resize, but reapply a new variable `out_flatten_level`?

File: paddle/operators/spp_op.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      StridedMemcpy<T>(context.device_context(),
+                       out_grad->data<T>() + out_offset, out_stride,
+                       outgrad_flatten_level.dims(), flatten_stride,
+                       outgrad_flatten_level.data<T>());
+      out_offset += out_flatten_level.dims()[1] * out_stride[1];
+      // flatten backward to nchw
+      framework::Tensor out_level;
+      framework::Tensor outgrad_level;
+      std::vector<int64_t> out_shape_vec({in_x->dims()[0], in_x->dims()[1]});
+      out_shape_vec.push_back(
```
Comment (NHZlX) [inline paddle/operators/spp_op.h]: ditto

File: paddle/operators/spp_op.cc (hunk 1 above)
Comment (sweetsky0901) [reply paddle/operators/spp_op.cc]: done

File: paddle/operators/spp_op.h (hunk 2 above)
Comment (sweetsky0901) [reply paddle/operators/spp_op.h]: done

File: paddle/operators/spp_op.h (hunk 3 above)
Comment (sweetsky0901) [reply paddle/operators/spp_op.h]: done

File: paddle/operators/spp_op.h (hunk 4 above)
Comment (sweetsky0901) [reply paddle/operators/spp_op.h]: done

**Comment by NHZlX** [comment]
`average` mode should be supported

**Pull request merged**